File: SourceCode/Python/General/GenTree.py
#/* *** ODSATag: GenTreeADT *** */
# General tree node ADT
class GTNode:
    def value(self): raise NotImplementedError
    def isLeaf(self): raise NotImplementedError
    def parent(self): raise NotImplementedError
    def leftmostChild(self): raise NotImplementedError
    def rightSibling(self): raise NotImplementedError
    def setValue(self, value): raise NotImplementedError
    def setParent(self, par): raise NotImplementedError
    def insertFirst(self, n): raise NotImplementedError
    def insertNext(self, n): raise NotImplementedError
    def removeFirst(self): raise NotImplementedError
    def removeNext(self): raise NotImplementedError

# General tree ADT
class GenTree:
    def clear(self): raise NotImplementedError
    def root(self): raise NotImplementedError
    def newroot(self, value, first, sib): raise NotImplementedError
    def newleftchild(self, value): raise NotImplementedError
}
#/* *** ODSAendTag: GenTreeADT *** */


#/* *** ODSATag: GenTreePrint *** */
# Preorder traversal for general trees
def preorder(rt):
    PrintNode(rt)
    if not rt.isLeaf():
        temp = rt.leftmostChild()
        while temp is not None
            preorder(temp)
            temp = temp.rightSibling()
#/* *** ODSAendTag: GenTreePrint *** */

File: SourceCode/Java_Generic/General/GenTree.java
/* *** ODSATag: GenTreeADT *** */
// General tree node ADT
public interface GTNode<E> {
    public E value();
    public boolean isLeaf();
    public GTNode<E> parent();
    public GTNode<E> leftmostChild();
    public GTNode<E> rightSibling();
    public void setValue(E value);
    public void setParent(GTNode<E> par);
    public void insertFirst(GTNode<E> n);
    public void insertNext(GTNode<E> n);
    public void removeFirst();
    public void removeNext();
}

// General tree ADT
public interface GenTree<E> {
    public void clear();       // Clear the tree
    public GTNode<E> root();   // Return the root
    // Make the tree have a new root, give first child and sib
    public void newRoot(E value, GTNode<E> first, GTNode<E> sib);
    public void newLeftChild(E value); // Add left child
}
/* *** ODSAendTag: GenTreeADT *** */


/* *** ODSATag: GenTreePrint *** */
// Preorder traversal for general trees
static void preorder(GTNode<E> node) {
    PrintNode(node);
    if (!node.isLeaf()) {
        GTNode<E> temp = node.leftmostChild();
        while (temp != null) {
            preorder(temp);
            temp = temp.rightSibling();
        }
    }
}
/* *** ODSAendTag: GenTreePrint *** */
